### ISSUE TITLE
Bugfix: Fixed MessageID not preserved when signing/encrypting with S/MIME and/or DKIM

### ISSUE THREAD

**Issue by honorhs** [bbottema/simple-java-mail#461]
hi.
I've found a serious bug in the DKIM authentication process.

This is the typical flow for sending mail

1. An internal messageId is generated and applied via `fixingMessageId()`
```
currentEmailBuilder.fixingMessageId("<123@456>");
```

2. DKIM configuration is performed for security enhancement
```
currentEmailBuilder.signWithDomainKey(
    DkimConfig.builder()
        .dkimPrivateKeyData(byte[] / File / InputStream)
        .dkimSigningDomain("your_domain.org")
        .dkimSelector("your_selector")
        .build()
);
```

3. The email is sent
```
mailer.sendMail(email);
```

---------

However, after step 3, the messageId value inside the email object is changed, even though the messageId was fixed in step 1.

The cause of the issue is the following line of code
[message.saveChanges()](https://github.com/bbottema/simple-java-mail/blob/82f8a1701fad71d203872cbf58ae61c429bd93dd/modules/simple-java-mail/src/main/java/org/simplejavamail/mailer/internal/SessionBasedEmailToMimeMessageConverter.java#L87)

Here, the messageId is changed by calling `message.saveChanges()`.

In any case, the messageId inside the email object should not be changed if it already exists.

There are two possible solutions

1. Restore the messageId inside the message object to its original value after calling `message.saveChanges()`.
2. If the messageId already exists, do not call `message.saveChanges()`. 
(There is no need to unnecessarily call saveChanges on the mimeMessage created by convertMimeMessage.)

The second solution seems to be the best

**Comment by bbottema** [comment]
Please provide evidence that this is not working. fixingMessageId is specifically made for this kind of situation. When message.saveChanges is called, the following override is triggered:

```java
MimeMessage message = new MimeMessage(session) {
			@Override
			protected void updateMessageID() throws MessagingException {
				if (valueNullOrEmpty(email.getId())) {
					super.updateMessageID();
				} else {
					setHeader("Message-ID", email.getId());
				}
			}
			
			@Override
			public String toString() {
				try {
					return format("MimeMessage<id:%s, subject:%s>", super.getMessageID(), super.getSubject());
				} catch (MessagingException e) {
					throw new IllegalStateException("should not reach here");
				}
			}
		};
```
This is also extensively automatically tested using a real (but for testing) SMTP server (in MailerLiveTest.java).

**Comment by honorhs** [comment]
I see the confusion you're having now.

Please take a look at the code a bit further down.

When returning through ModuleLoader, MimeMessage is overridden again. This means that when using features such as Smime or Dkim, MimeMessage is overridden again. [code](https://github.com/bbottema/simple-java-mail/blob/82f8a1701fad71d203872cbf58ae61c429bd93dd/modules/simple-java-mail/src/main/java/org/simplejavamail/converter/internal/mimemessage/SpecializedMimeMessageProducer.java#L84-L99)

```java
if (email.getPkcs12ConfigForSmimeSigning() != null) {
	message = ModuleLoader.loadSmimeModule().signMessageWithSmime(session, message, email.getPkcs12ConfigForSmimeSigning());
}

if (email.getX509CertificateForSmimeEncryption() != null) {
	message = ModuleLoader.loadSmimeModule().encryptMessageWithSmime(session, message, email.getX509CertificateForSmimeEncryption());
}

if (email.getDkimConfig() != null) {
	message = ModuleLoader.loadDKIMModule().signMessageWithDKIM(message, email.getDkimConfig(), checkNonEmptyArgument(email.getFromRecipient(), "fromRecipient"));
}

if (email.getBounceToRecipient() != null) {
	// display name not applicable: https://tools.ietf.org/html/rfc5321#section-4.1.2
	message = new ImmutableDelegatingSMTPMessage(message, email.getBounceToRecipient().getAddress());
}
```


In other words, if use the Dkim feature, the overriding of the updateMessageID method above becomes useless.

**Comment by honorhs** [comment]
[DKIMSigner. signMessageWithDKIM(...)](https://github.com/bbottema/simple-java-mail/blob/82f8a1701fad71d203872cbf58ae61c429bd93dd/modules/dkim-module/src/main/java/org/simplejavamail/internal/dkimsupport/DKIMSigner.java#L54)
```java
....
return new DkimMessage(messageToSign, dkimSigner);
```

DkimMessage (in utils-mail-dkim-3.0.0.jar)
```java
public class DkimMessage extends SMTPMessage {

	private static byte[] NL = { (byte) '\r', (byte) '\n' };

	private DkimSigner signer;

	private String encodedBody;

	/**
	 * Created a new {@code DkimMessage} from the given {@link MimeMessage} and
	 * {@link DkimSigner}.
	 * 
	 * @param message
	 *            The {@link MimeMessage} to be signed.
	 * @param signer
	 *            The {@link DkimSigner} to sign the message with.
	 * @throws MessagingException
	 *             If constructing this {@code DkimMessage} failed.
	 */
	public DkimMessage(MimeMessage message, DkimSigner signer) throws MessagingException {
		super(message);
		this.signer = signer;
	}
	...
```


DkimMessage is the result of extending Jakarta MimeMessage.

**Comment by bbottema** [comment]
Oh wauw, that's a nice find! I'll see about reproducing this in my tests so that's covered. I'll fix this, but I'm not sure how fast. Might be a week or two.

Thank you for bringing this to my attentions.

**Comment by honorhs** [comment]
Thank you for taking the time to fix the bug. 
I hope to see the issue resolved quickly. 👍

**Comment by honorhs** [comment]
@bbottema 
Are there any updates?

**Comment by bbottema** [comment]
Fix released in v8.1.1. Can you please verify?

**Comment by kkrupka** [comment]
I have just updated to 8.1.1 (from 8.0.0) but it's probably related to this ticket as the mentioned class was introduced in the above mentioned commit. It looks like anything went wrong, because I get a ClassNotFoundException:

```
Caused by: java.lang.NoClassDefFoundError: org/simplejavamail/internal/dkimsupport/DkimMessageIdFixingMimeMessage
	at org.simplejavamail.mailer.internal.SessionBasedEmailToMimeMessageConverter.convertAndLogMimeMessage(SessionBasedEmailToMimeMessageConverter.java:108)
	at org.simplejavamail.mailer.internal.SessionBasedEmailToMimeMessageConverter.convertAndLogMimeMessage(SessionBasedEmailToMimeMessageConverter.java:81)
	at org.simplejavamail.mailer.internal.util.TransportRunner.lambda$sendMessage$0(TransportRunner.java:52)
	at org.simplejavamail.mailer.internal.util.TransportRunner.sendUsingConnectionPool(TransportRunner.java:84)
	at org.simplejavamail.mailer.internal.util.TransportRunner.runOnSessionTransport(TransportRunner.java:69)
	at org.simplejavamail.mailer.internal.util.TransportRunner.sendMessage(TransportRunner.java:51)
	at org.simplejavamail.mailer.internal.SendMailClosure.executeClosure(SendMailClosure.java:69)
	at org.simplejavamail.mailer.internal.AbstractProxyServerSyncingClosure.run(AbstractProxyServerSyncingClosure.java:56)
	at org.simplejavamail.mailer.internal.MailerImpl.sendMail(MailerImpl.java:363)
	at org.simplejavamail.mailer.internal.MailerImpl.sendMail(MailerImpl.java:348)
[...cropped...]
Caused by: java.lang.ClassNotFoundException: org.simplejavamail.internal.dkimsupport.DkimMessageIdFixingMimeMessage
```
Under 8.0.0 it was working and I got no exception. Without SMIME signing information everyhting works. Do you have an idea?

**Comment by kkrupka** [comment]
Just if someone else experience a `java.lang.ClassNotFoundException: org.simplejavamail.internal.dkimsupport.DkimMessageIdFixingMimeMessage`. Up to now, I used nothing from the dkim-module dependency and therefore haven't listed it in my pom. But with this update the dkim-module is required for the runtime otherwise the above exception is thrown when sending signed emails with S/MIME. Error does not occur on compile time.

**Comment by bbottema** [comment]
Yikes, you're right! I already know why it happens, but a quick work around is indeed to have the dkim module on the classpath. I'll send out a fix.

**Comment by bbottema** [comment]
> Just if someone else experience a `java.lang.ClassNotFoundException: org.simplejavamail.internal.dkimsupport.DkimMessageIdFixingMimeMessage`. Up to now, I used nothing from the dkim-module dependency and therefore haven't listed it in my pom. But with this update the dkim-module is required for the runtime otherwise the above exception is thrown when sending signed emails with S/MIME. Error does not occur on compile time.

Fix released in v8.1.2!

**Comment by kkrupka** [comment]
You are great. Super quick fix! I can confirm, it works as before.
